tests/contracts/wallet_2.scilla

```diff
@@ -0,0 +1,392 @@
+scilla_version 0
+
+import ListUtils IntUtils BoolUtils
+
+(***************************************************)
+(*               Associated library                *)
+(***************************************************)
+library WalletLib
+
+(* Event emitted when the contract is initialized *)
+let mk_contract_initialized_event =
+  { _eventname : "Contract initialized" }
+
+(* Event for communicating a new transaction id *)
+let mk_transaction_added_event =
+  fun (tc : Uint32) =>
+    { _eventname : "Transaction created" ; transactionId : tc }
+
+(* Event for communicating the signing of a transaction *)
+let mk_transaction_signed_event =
+    { _eventname : "Transaction signed" }
+
+type Error =
+| NonOwnerCannotSign
+| UnknownTransactionId
+| InsufficientFunds
+| NoSignatureListFound
+| AlreadySigned
+| NotAlreadySigned
+| InvalidContract
+| InvalidAmount
+| NotEnoughSignatures
+| SenderMayNotExecute
+| NonOwnerCannotSubmit
+
+(* Error events *)
+let mk_error_event =
+  fun (err : Error) =>
+  let err_code = 
+    match err with
+    | NonOwnerCannotSign    => Int32 -1
+    | UnknownTransactionId  => Int32 -2
+    | InsufficientFunds     => Int32 -3
+    | NoSignatureListFound  => Int32 -4
+    | AlreadySigned         => Int32 -5
+    | NotAlreadySigned      => Int32 -6
+    | InvalidContract       => Int32 -7
+    | InvalidAmount         => Int32 -8
+    | NotEnoughSignatures   => Int32 -9
+    | SenderMayNotExecute   => Int32 -10
+    | NonOwnerCannotSubmit  => Int32 -11
+    end in
+  { _eventname : "WalletError" ; err_code : err_code }
+
+let transaction_inc = Uint32 1
+let t = True
+
+(* One (potential) transaction, consisting of a recipient address, an amount, *)
+(* and a tag (in case the recipient is another contract *)
+type Transaction =
+| Trans of ByStr20 Uint128 String
+
+(* Make map of owners *)
+let mk_owners_map =
+  fun (owners : List ByStr20) =>
+    let init = Emp ByStr20 Bool in
+    let iter =
+      fun (acc : Map ByStr20 Bool) =>
+      fun (cur_owner : ByStr20) =>
+        let mem = builtin get acc cur_owner in
+        match mem with
+        | Some True =>
+          (* owner already added *)
```

Comment:
Perhaps this should be a part of the contract's validity check. The presence of duplicate entries in the owners list would likely indicate e.g. a copy-paste error.

Reply:
The owners map is now built as part of the validity check (guarded by the other checks, so that we don't build it if something else is wrong).

Just need to change the name of the `check_validity` function.

Reply:
In `mk_owners_map` (here), we don't need to check if the owner is already added right? we can blindly add it, and rely on the size based check later to see if there was a duplicate (which also is already written).

In other words, we can just do `builtin put acc cur_owner t` without `get`ing and checking for membership.

Reply:
Good point. I'll change it.